src/api-binder.ts

```diff
@@ -0,0 +1,890 @@
+			},
+			this.cachedBalenaApi.passthrough,
+		);
+
+		await this.cachedBalenaApi._request(requestParams);
+	}
+
+	private async report() {
+		const conf = await this.config.getMany([
+			'deviceId',
+			'apiTimeout',
+			'apiEndpoint',
+			'uuid',
+			'localMode',
+		]);
+
+		if (checkTruthy(conf.localMode || false)) {
+			return;
+		}
+
+		const stateDiff = this.getStateDiff();
+		if (_.size(stateDiff) === 0) {
+			return 0;
+		}
+
+		await Bluebird.resolve(
+			this.sendReportPatch(stateDiff, conf as {
+				uuid: string;
+				apiEndpoint: string;
+			}),
+		).timeout(conf.apiTimeout as number);
+
+		this.stateReportErrors = 0;
+		_.assign(this.lastReportedState.local, stateDiff.local);
+		_.assign(this.lastReportedState.dependent, stateDiff.dependent);
+	}
+
+	private reportCurrentState(): null {
+		this.reportPending = true;
+		// TODO: Once device-state has been converted to typescript,
+		// remove this any
+		this.deviceState
+			.getStatus()
+			.then((currentDeviceState: any) => {
+				_.assign(this.stateForReport.local, currentDeviceState.local);
+				_.assign(this.stateForReport.dependent, currentDeviceState.dependent);
+
+				const stateDiff = this.getStateDiff();
+				if (_.size(stateDiff) === 0) {
+					this.reportPending = false;
+					return null;
+				}
+				return Bluebird.resolve(this.report())
+					.delay(REPORT_SUCCESS_DELAY)
+					.then(() => {
+						this.reportCurrentState();
+					});
+			})
+			.catch((e: Error) => {
+				this.eventTracker.track('Device state report failure', { error: e });
+				const delay = Math.min(
+					2 ** this.stateReportErrors * 500,
+					MAX_REPORT_RETRY_DELAY,
+				);
+
+				++this.stateReportErrors;
+				return Bluebird.delay(delay).then(this.reportCurrentState);
+			});
+		return null;
+	}
+
+	private getAndSetTargetState(force: boolean, isFromApi = false) {
+		return Bluebird.using(this.lockGetTarget(), async () => {
+			const targetState = await this.getTargetState();
+			if (isFromApi || !_.isEqual(targetState, this.lastTarget)) {
+				await this.deviceState.setTarget(targetState);
+				this.lastTarget = _.cloneDeep(targetState);
+				this.deviceState.triggerApplyTarget({ force });
+			}
+		})
+			.tapCatch(err => {
+				console.error(`Failed to get target state for device: ${err}`);
+			})
+			.finally(() => {
+				this.lastTargetStateFetch = process.hrtime();
+			});
+	}
+
+	private async pollTargetState(): Promise<void> {
+		let pollInterval = await this.config.get('appUpdatePollInterval');
+		if (!_.isNumber(pollInterval)) {
+			throw new InternalInconsistencyError(
+				'appUpdatePollInterval not a number in ApiBinder.pollTargetState',
+			);
+		}
+
+		try {
+			await this.getAndSetTargetState(false);
+			this.targetStateFetchErrors = 0;
+		} catch (e) {
+			pollInterval = Math.min(
+				pollInterval,
+				15000 * 2 ** this.targetStateFetchErrors,
+			);
+			++this.targetStateFetchErrors;
+		}
+
+		return Bluebird.delay(pollInterval).then(this.pollTargetState);
```

Comment:
This can be
```suggestion
		await Bluebird.delay(pollInterval);
		return this.pollTargetState()
```
to keep the style more consistent